### PR TITLE
Separate GNU TLS extension from websockets

### PR DESCRIPTION
**Pull request by GiedriusM**
Previously the only way to use client certificate was to enable websockets. However due to libmicrohttpd version requirement it could not be done without a hassle in certain setups. This change allows using HTTPS client certificate without websocket (and libmicrohttpd >=0.9.53).

Some notes:
* Currently GNU TLS extension "documentation" is added into "What's new in ulfius 2.0" section. Most likely it should be placed somewhere else, just wasn't sure where exactly.
* Enabling websockets will force-enable GNU TLS extension. I was reluctant to do complete GNU TLS separation from websockets since I don't use them and could've easily broke something without noticing.
* Makefiles in the example_programs folder use ` GNUTLSFLAG=-DU_DISABLE_GNUTLS` just for consistency. Using ` GNUTLSFLAG=1` is sufficient.
* Changes in `example_programs/cmake-modules/FindUlfius.cmake` are untested (copy/paste/replace).

### CODE REVIEW

**Comment by GiedriusM** [comment]
Looks like travis libgnutls version is <3.1.3 and `gnutls_x509_crt_export2` is not available. Should I change it to `gnutls_x509_crt_export` or set travis to `xenial`?

**Comment by babelouest** [comment]
I have to admit, build options can be inconsistent and hide functionalities behind dependencies version or the opposite.

Correct me if I'm wrong, but what you need is to send https requests and use client cert authentification, with an "old" version of libmicrohttpd (< 0.9.53), so without websocket service support.

So you could build ulfius with a command like:

```shell
$ cmake -DWITH_JANSSON=ON -DWITH_CURL=ON -DWITH_WEBSOCKET=OFF -DBUILD_UWSC=OFF -DWITH_GNUTLS=ON ..
```

Is that so?

This is a good opportunity to rethink a little bit how options are entangled and dependent.

For example, setting `-DWITH_GNUTLS=OFF` will imply disabling websockets and uwsc, even if they are set to `ON`.

As you mentionned, we should also fix travis script to be able to test most options as possible.

File: API.md [hunk 1]
```diff
@@ -1563,11 +1563,29 @@ If you need to communicate between callback functions for any purpose, you can u
 
 the values `string_body` and `json_body` have been removed from the structures `struct _u_request` and `struct _u_response`. This may be painless in the response if you used only the functions `ulfius_set_xxx_body_response`. Otherwise, you should make small arrangements to your code.
 
+### GNU TLS extension
+Ulfius GNU TLS extension allows sending HTTPS requests with client certificate and enables additional client certificate validation option for HTTS server.
```
Comment (babelouest) [inline API.md]: > option for HTTS server.

=> `option for HTTPS server.`

File: API.md [hunk 2]
```diff
@@ -1563,11 +1563,29 @@ If you need to communicate between callback functions for any purpose, you can u
...
+
+Using websocket requires [libgnutls](https://www.gnutls.org/).
+
+This extension can be disabled with `GNUTLSFLAG` option when using make:
+
+```shell
+$ make GNUTLS=1 WEBSOCKETFLAG=1
```
Comment (babelouest) [inline API.md]: I'd use the `INSTALL.md` to explain Makefile options, I prefer `API.md` to be read after installation.

File: example_programs/injection_example/Makefile [hunk 3]
```diff
@@ -33,7 +33,7 @@ debug: ADDITIONALFLAGS=-DDEBUG -g -O0
 debug: injection_example
 
 ../../src/libulfius.so:
-	cd $(ULFIUS_LOCATION) && $(MAKE) debug JANSSONFLAG=-DU_DISABLE_JANSSON CURLFLAG=-DU_DISABLE_CURL WEBSOCKETFLAG=-DU_DISABLE_WEBSOCKET
+	cd $(ULFIUS_LOCATION) && $(MAKE) debug JANSSONFLAG=-DU_DISABLE_JANSSON CURLFLAG=-DU_DISABLE_CURL WEBSOCKETFLAG=-DU_DISABLE_WEBSOCKET GNUTLSFLAG=-DU_DISABLE_GNUTLS
```
Comment (babelouest) [inline example_programs/injection_example/Makefile]: even better:
```shell
cd $(ULFIUS_LOCATION) && $(MAKE) debug JANSSONFLAG=-DU_DISABLE_JANSSON CURLFLAG=-DU_DISABLE_CURL GNUTLSFLAG=-DU_DISABLE_GNUTLS
```

File: example_programs/multiple_callbacks_example/Makefile [hunk 4]
```diff
@@ -33,7 +33,7 @@ debug: ADDITIONALFLAGS=-DDEBUG -g -O0
 debug: multiple_callbacks_example
 
 ../../src/libulfius.so:
-	cd $(ULFIUS_LOCATION) && $(MAKE) debug JANSSONFLAG=-DU_DISABLE_JANSSON CURLFLAG=-DU_DISABLE_CURL WEBSOCKETFLAG=-DU_DISABLE_WEBSOCKET
+	cd $(ULFIUS_LOCATION) && $(MAKE) debug JANSSONFLAG=-DU_DISABLE_JANSSON CURLFLAG=-DU_DISABLE_CURL WEBSOCKETFLAG=-DU_DISABLE_WEBSOCKET GNUTLSFLAG=-DU_DISABLE_GNUTLS
```
Comment (babelouest) [inline example_programs/multiple_callbacks_example/Makefile]:
```shell
cd $(ULFIUS_LOCATION) && $(MAKE) debug JANSSONFLAG=-DU_DISABLE_JANSSON CURLFLAG=-DU_DISABLE_CURL GNUTLSFLAG=-DU_DISABLE_GNUTLS
```

File: example_programs/proxy_example/Makefile [hunk 5]
```diff
@@ -33,7 +33,7 @@ debug: ADDITIONALFLAGS=-DDEBUG -g -O0
 debug: proxy
 
 ../../src/libulfius.so:
-	cd $(ULFIUS_LOCATION) && $(MAKE) debug JANSSONFLAG=-DU_DISABLE_JANSSON WEBSOCKETFLAG=-DU_DISABLE_WEBSOCKET
+	cd $(ULFIUS_LOCATION) && $(MAKE) debug JANSSONFLAG=-DU_DISABLE_JANSSON WEBSOCKETFLAG=-DU_DISABLE_WEBSOCKET GNUTLSFLAG=-DU_DISABLE_GNUTLS
```
Comment (babelouest) [inline example_programs/proxy_example/Makefile]:
```shell
cd $(ULFIUS_LOCATION) && $(MAKE) debug JANSSONFLAG=-DU_DISABLE_JANSSON GNUTLSFLAG=-DU_DISABLE_GNUTLS
```

File: example_programs/request_example/Makefile [hunk 6]
```diff
@@ -33,7 +33,7 @@ debug: ADDITIONALFLAGS=-DDEBUG -g -O0
 debug: client server mail
 
 ../../src/libulfius.so:
-	cd $(ULFIUS_LOCATION) && $(MAKE) debug WEBSOCKETFLAG=-DU_DISABLE_WEBSOCKET
+	cd $(ULFIUS_LOCATION) && $(MAKE) debug WEBSOCKETFLAG=-DU_DISABLE_WEBSOCKET GNUTLSFLAG=-DU_DISABLE_GNUTLS
```
Comment (babelouest) [inline example_programs/request_example/Makefile]:
```shell
cd $(ULFIUS_LOCATION) && $(MAKE) debug GNUTLSFLAG=-DU_DISABLE_GNUTLS
```

File: example_programs/sheep_counter/Makefile [hunk 7]
```diff
@@ -33,7 +33,7 @@ debug: ADDITIONALFLAGS=-DDEBUG -g -O0
 debug: sheep_counter
 
 ../../src/libulfius.so:
-	cd $(ULFIUS_LOCATION) && $(MAKE) debug CURLFLAG=-DU_DISABLE_CURL WEBSOCKETFLAG=-DU_DISABLE_WEBSOCKET
+	cd $(ULFIUS_LOCATION) && $(MAKE) debug CURLFLAG=-DU_DISABLE_CURL WEBSOCKETFLAG=-DU_DISABLE_WEBSOCKET GNUTLSFLAG=-DU_DISABLE_GNUTLS
```
Comment (babelouest) [inline example_programs/sheep_counter/Makefile]:
```shell
cd $(ULFIUS_LOCATION) && $(MAKE) debug CURLFLAG=-DU_DISABLE_CURL GNUTLSFLAG=-DU_DISABLE_GNUTLS
```

File: example_programs/simple_example/Makefile [hunk 8]
```diff
@@ -34,7 +34,7 @@ debug: ADDITIONALFLAGS=-DDEBUG -g -O0
 debug: simple_example
 
 ../../src/libulfius.so:
-	cd $(ULFIUS_LOCATION) && $(MAKE) debug CURLFLAG=-DU_DISABLE_CURL WEBSOCKETFLAG=-DU_DISABLE_WEBSOCKET
+	cd $(ULFIUS_LOCATION) && $(MAKE) debug CURLFLAG=-DU_DISABLE_CURL WEBSOCKETFLAG=-DU_DISABLE_WEBSOCKET GNUTLSFLAG=-DU_DISABLE_GNUTLS
```
Comment (babelouest) [inline example_programs/simple_example/Makefile]:
```shell
cd $(ULFIUS_LOCATION) && $(MAKE) debug CURLFLAG=-DU_DISABLE_CURL GNUTLSFLAG=-DU_DISABLE_GNUTLS
```

File: example_programs/stream_example/Makefile [hunk 9]
```diff
@@ -33,7 +33,7 @@ debug: ADDITIONALFLAGS=-DDEBUG -g -O0
 debug: stream_example stream_client
 
 ../../src/libulfius.so:
-	cd $(ULFIUS_LOCATION) && $(MAKE) debug JANSSONFLAG=-DU_DISABLE_JANSSON WEBSOCKETFLAG=-DU_DISABLE_WEBSOCKET
+	cd $(ULFIUS_LOCATION) && $(MAKE) debug JANSSONFLAG=-DU_DISABLE_JANSSON WEBSOCKETFLAG=-DU_DISABLE_WEBSOCKET GNUTLSFLAG=-DU_DISABLE_GNUTLS
```
Comment (babelouest) [inline example_programs/stream_example/Makefile]:
```shell
cd $(ULFIUS_LOCATION) && $(MAKE) debug JANSSONFLAG=-DU_DISABLE_JANSSON GNUTLSFLAG=-DU_DISABLE_GNUTLS
```

File: example_programs/test_u_map/Makefile [hunk 10]
```diff
@@ -33,7 +33,7 @@ debug: ADDITIONALFLAGS=-DDEBUG -g -O0
 debug: test_u_map
 
 ../../src/libulfius.so:
-	cd $(ULFIUS_LOCATION) && $(MAKE) debug JANSSONFLAG=-DU_DISABLE_JANSSON CURLFLAG=-DU_DISABLE_CURL WEBSOCKETFLAG=-DU_DISABLE_WEBSOCKET
+	cd $(ULFIUS_LOCATION) && $(MAKE) debug JANSSONFLAG=-DU_DISABLE_JANSSON CURLFLAG=-DU_DISABLE_CURL WEBSOCKETFLAG=-DU_DISABLE_WEBSOCKET GNUTLSFLAG=-DU_DISABLE_GNUTLS
```
Comment (babelouest) [inline example_programs/test_u_map/Makefile]:
```shell
cd $(ULFIUS_LOCATION) && $(MAKE) debug JANSSONFLAG=-DU_DISABLE_JANSSON CURLFLAG=-DU_DISABLE_CURL GNUTLSFLAG=-DU_DISABLE_GNUTLS
```

**Comment by GiedriusM** [comment]
> Correct me if I'm wrong, but what you need is to send https requests and use client cert authentification, with an "old" version of libmicrohttpd (< 0.9.53), so without websocket service support.

Exactly.

> For example, setting `-DWITH_GNUTLS=OFF` will imply disabling websockets and uwsc, even if they are set to `ON`.

Currently setting `-DWITH_WEBSOCKETS=ON` (default) will enable GNU TLS, but it could be done the other around I guess. Matter of preference.
Anyway, in theory websockets can work on plain HTTP and do not require HTTPS, therefore this is not a must, i.e. both `cmake .. -DWITH_WEBSOCKET=on -DWITH_GNUTLS=on` and `cmake .. -WITH_WEBSOCKET=on -DWITH_GNUTLS=off` could be two valid and different configurations. I could do the separation in this PR, but would need someone to test it or give advice for test scenarious as I'm not using websockets and therefore am not too familiar.

File: src/Makefile [hunk 11]
```diff
@@ -46,11 +46,18 @@ endif
...
 DISABLE_WEBSOCKET=0
-LWEBSOCKET=-lgnutls
+GNUTLSFLAG=1
 else
 DISABLE_WEBSOCKET=1
 endif
 
+ifndef GNUTLSFLAG
+DISABLE_GNUTLS=0
+LWEBSOCKET=-lgnutls
```
Comment (babelouest) [inline src/Makefile]: In that case, I'd rename this flag `LGNUTLS`

File: CMakeLists.txt [hunk 12]
```diff
@@ -113,6 +108,29 @@ else ()
     set(U_DISABLE_WEBSOCKET ON)
 endif ()
 
+# GNU TLS support
+option(WITH_GNUTLS "GNU TLS support" ON)
+
+# In theory websockets do not require TLS.
```
Comment (babelouest) [inline CMakeLists.txt]: No but websocket requires SHA1 encoding which is provided by gnutls.

**Comment by babelouest** [comment]
I could make travis ci script work with ubuntu xenial and gnutls. I'll make new tests this week-end to merge [your branch](https://github.com/babelouest/ulfius/tree/punica-pr-separate-gnutls) in the master.

Thanks for the pull request @GiedriusM !

**Comment by GiedriusM** [comment]
No problem.

Also I've noticed that websockets are disabled in windows. Is this due to gnutls, libmicrohttpd or some other reason? If it's because of libgnutls, then I should slightly update the docs and the makefiles.

**Comment by GiedriusM** [comment]
BTW, an alternative to xenial could be https://github.com/punica/ulfius/pull/1/commits/0f6ef9c7ff4398a6802ccd412cc600c50ea66dfb.

**Comment by babelouest** [comment]
Ulfius doesn't have websocket on Windows because there were problems with the `poll()` function.

Concerning the patch you mention, that's interesting, but what platform and GnuTLS version are you using?
Because according to [GnuTLS documentation](https://gnutls.org/reference/gnutls-x509.html#gnutls-x509-crt-export2), `gnutls_x509_crt_export2` exists since GnuTLS 3.1.3, which was released in 2012.

**Comment by GiedriusM** [comment]
> Ulfius doesn't have websocket on Windows because there were problems with the `poll()` function.

Alright, so it's all good then.

> Concerning the patch you mention, that's interesting, but what platform and GnuTLS version are you using?
> Because according to [GnuTLS documentation](https://gnutls.org/reference/gnutls-x509.html#gnutls-x509-crt-export2), `gnutls_x509_crt_export2` exists since GnuTLS 3.1.3, which was released in 2012.

It's not so much of a problem for me as it is for Travis :) By default it's using Ubuntu Trusty, which installs  libgnutls-dev 2.12.23 (https://packages.ubuntu.com/trusty/libgnutls-dev). Starting from Xenial it's all good, though. Just wanted to give an alternative, which has a negligible advantage of working on (soon to be) non-LTS Ubuntu versions.

**Comment by babelouest** [comment]
> It's not so much of a problem for me as it is for Travis :) By default it's using Ubuntu Trusty, which installs libgnutls-dev 2.12.23 (https://packages.ubuntu.com/trusty/libgnutls-dev). Starting from Xenial it's all good, though. Just wanted to give an alternative, which has a negligible advantage of working on (soon to be) non-LTS Ubuntu versions.

It's settled for GnuTLS then, thanks to you I will use ubuntu xenial in Travis CI.
But I will not work on compatibility with old libraries versions as is, I prefer focusing on new features and maintaining the library stable with the type of environment most users have.

**Comment by babelouest** [comment]
@GiedriusM , Ulfius 2.5.3 has just been released with your changes in it, happy hacking!

**Comment by GiedriusM** [comment]
@babelouest, thanks!

**Pull request merged**